frontend-react/src/components/FileHandlers/FileHandler.tsx

```diff
@@ -0,0 +1,309 @@
+import React, { useState, useMemo } from "react";
+
+import { showError } from "../AlertNotifications";
+import { useSessionContext } from "../../contexts/SessionContext";
+import { useOrganizationResource } from "../../hooks/UseOrganizationResouce";
+import { FileResponseError } from "../../network/api/WatersApi";
+import { WatersPost } from "../../network/api/WatersApiFunctions";
+import { Destination } from "../../resources/ActionDetailsResource";
+import Spinner from "../Spinner"; // TODO: refactor to use suspense
+
+import { FileErrorDisplay, FileSuccessDisplay } from "./FileHandlerMessaging";
+import { FileHandlerForm } from "./FileHandlerForm";
+
+// values taken from Report.kt
+const PAYLOAD_MAX_BYTES = 50 * 1000 * 1000; // no idea why this isn't in "k" (* 1024).
+const REPORT_MAX_ITEMS = 10000;
+const REPORT_MAX_ITEM_COLUMNS = 2000;
+
+const parseCsvForError = (fileName: string, filecontent: string) => {
+    const error = false;
+    // count the number of lines
+    const linecount = (filecontent.match(/\n/g) || []).length + 1;
+    if (linecount > REPORT_MAX_ITEMS) {
+        showError(
+            `The file '${fileName}' has too many rows. The maximum number of rows allowed is ${REPORT_MAX_ITEMS}.`
+        );
+        return true;
+    }
+    if (linecount <= 1) {
+        showError(
+            `The file '${fileName}' doesn't contain any valid data. ` +
+                `File should have a header line and at least one line of data.`
+        );
+        return true;
+    }
+
+    // get the first line and examine it
+    const firstline = (filecontent.match(/^(.*)\n/) || [""])[0];
+    // ideally, the columns would be comma seperated, but they may be tabs, because the first
+    // line is a header, we don't have to worry about escaped delimiters in strings (e.g. ,"Smith, John",)
+    const columncount =
+        (firstline.match(/,/g) || []).length ||
+        (firstline.match(/\t/g) || []).length;
+
+    if (columncount > REPORT_MAX_ITEM_COLUMNS) {
+        showError(
+            `The file '${fileName}' has too many columns. The maximum number of allowed columns is ${REPORT_MAX_ITEM_COLUMNS}.`
+        );
+        return true;
+    }
+    // todo: this is a good place to do basic validation of the upload file. e.g. does it have
+    // all the required columns? Are any rows obviously not correct (empty or obviously wrong type)?
+    return error;
+};
+
+interface FileHandlerProps {
```

Comment:
the key here is that this component, unlike the existing `Validate` can be customized via these props